src/functional/ffront/foast_passes/type_deduction.py

```diff
@@ -664,6 +664,42 @@ def _visit_max_over(self, node: foast.Call, **kwargs) -> foast.Call:
     def _visit_min_over(self, node: foast.Call, **kwargs) -> foast.Call:
         return self._visit_reduction(node, **kwargs)
 
+    def _visit_astype(self, node: foast.Call, **kwargs) -> foast.Call:
+        casted_obj_type = node.args[0].type
+        dtype_obj = node.args[1]
+        assert isinstance(dtype_obj, foast.Name)
+        dtype_obj_type = dtype_obj.type
+        assert isinstance(dtype_obj_type, ct.FunctionType)
+        dtype_obj_type_returns = dtype_obj_type.returns
+        if dtype_obj.id not in fbuiltins.TYPE_BUILTIN_NAMES or not (
+            type_info.is_arithmetic(dtype_obj_type_returns)
+            or type_info.is_logical(dtype_obj_type_returns)
+        ):
+            raise FieldOperatorTypeDeductionError.from_foast_node(
+                node,
+                msg=f"Incompatible argument in call to `{node.func.id}`'s new dtype. "
+                f"Expected an arithmetic or boolean dtype, but got {dtype_obj_type}.",
+            )
+        if not isinstance(casted_obj_type, ct.FieldType) or not (
+            type_info.is_arithmetic(casted_obj_type) or type_info.is_logical(casted_obj_type)
+        ):
+            raise FieldOperatorTypeDeductionError.from_foast_node(
+                node,
+                msg=f"Incompatible argument in call to `{node.func.id}`'s object. "
+                f"Expected arithmetic or boolean FieldType, but got {casted_obj_type}.",
+            )
```

Comment:
maybe should be assert as `as_type` already guarantees that the first argument is a FieldType